File: src-tauri/src/queue_list.rs
/*
 * Flex Flow
 * Copyright (c) 2025-present fdar0536
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

use crate::loadlib::api;
use crate::ffmodeldef as def;

#[tauri::command]
pub fn queue_list_init(connect_handle: def::Handle) -> Result<def::Handle, u8>
{
    let mut out = 0u32;
    let ret = (api().queue_list.init)(connect_handle, &mut out);
    if ret == 0 { Ok(out) } else { Err(ret) }
}

#[tauri::command]
pub fn queue_list_destroy(h: def::Handle) -> Result<(), u8>
{
    let ret = (api().queue_list.destroy)(h);
    if ret == 0 { Ok(()) } else { Err(ret) }
}

#[tauri::command]
pub async fn queue_list_create_queue(h: def::Handle, name: &str)
-> Result<(), u8>
{
    let name = name.to_owned();
    return tokio::task::spawn_blocking(move ||
    {
        let c_name = std::ffi::CString::new(name).map_err(|_| 255u8)?;
        let ret = (api().queue_list.create_queue)(h, c_name.as_ptr());
        if ret == 0 { Ok(()) } else { Err(ret) }
    })
    .await
    .unwrap_or(Err(255));
}

#[tauri::command]
pub async fn queue_list_list_queue(h: def::Handle) -> Result<Vec<String>, u8>
{
    return tokio::task::spawn_blocking(move ||
    {
        let mut out_ptr: *mut *mut i8 = std::ptr::null_mut();
        let mut out_size: usize = 0;

        let ret = (api().queue_list.list_queue)
        (h, &mut out_ptr as *mut *mut *mut i8, &mut out_size);
        if ret != 0
        {
            return Err(ret);
        }

        let slice =
        unsafe { std::slice::from_raw_parts(out_ptr, out_size) };
        let mut result = Vec::with_capacity(out_size);

        for &ptr in slice {
            let cstr = unsafe { std::ffi::CStr::from_ptr(ptr) } ;
            let s = cstr.to_string_lossy().into_owned();
            result.push(s);

            unsafe { libc::free(ptr as *mut libc::c_void) };
        }

        unsafe { libc::free(out_ptr as *mut libc::c_void) };

        Ok(result)
    })
    .await
    .unwrap_or(Err(255));
}

#[tauri::command]
pub async fn queue_list_delete_queue(h: def::Handle, name: &str)
-> Result<(), u8>
{
    let name = name.to_owned();
    return tokio::task::spawn_blocking(move ||
    {
        let c_name = std::ffi::CString::new(name).map_err(|_| 255u8)?;
        let ret = (api().queue_list.delete_queue)(h, c_name.as_ptr());
        if ret == 0 { Ok(()) } else { Err(ret) }
    })
    .await
    .unwrap_or(Err(255));
}

#[tauri::command]
pub async fn queue_list_rename_queue
(h: def::Handle, old: &str, new: &str) -> Result<(), u8>
{
    let old = old.to_owned();
    let new = new.to_owned();
    return tokio::task::spawn_blocking(move ||
    {
        let c_old = std::ffi::CString::new(old).map_err(|_| 255u8)?;
        let c_new = std::ffi::CString::new(new).map_err(|_| 255u8)?;
        let ret = (api().queue_list.rename_queue)
        (h, c_old.as_ptr(), c_new.as_ptr());
        if ret == 0 { Ok(()) } else { Err(ret) }
    })
    .await
    .unwrap_or(Err(255));
}

#[tauri::command]
pub fn queue_list_get_queue(h: def::Handle, name: &str)
-> Result<def::Handle, u8>
{
    let c_name = std::ffi::CString::new(name).map_err(|_| 255u8)?;
    let mut out = 0u32;
    let ret = (api().queue_list.get_queue)(h, c_name.as_ptr(), &mut out);
    if ret == 0 { Ok(out) } else { Err(ret) }
}

#[tauri::command]
pub fn queue_list_return_queue
(h: def::Handle, queue: def::Handle) -> Result<(), u8>
{
    let ret = (api().queue_list.return_queue)(h, queue);
    if ret == 0 { Ok(()) } else { Err(ret) }
}

File: src-cpp/model/dao/grpc/queuelist.cpp
/*
 * Flex Flow
 * Copyright (c) 2023-2024 fdar0536
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "spdlog/spdlog.h"

#include "controller/global/global.hpp"
#include "model/errmsg.hpp"

#include "connect.hpp"
#include "queue.hpp"
#include "queuelist.hpp"
#include "utils.hpp"

namespace Model
{

namespace DAO
{

namespace GRPC
{

QueueList::QueueList() :
    m_stub(nullptr)
{}

QueueList::~QueueList()
{
    if (m_conn) delete m_conn;
}

u8 QueueList::init(IConnect *connect)
{
    spdlog::debug("{}:{} QueueList::init", LOG_FILE_PATH(__FILE__), __LINE__);

    if (!connect)
    {
        spdlog::error("{}:{} connect is nullptr", LOG_FILE_PATH(__FILE__), __LINE__);
        return ErrCode_INVALID_ARGUMENT;
    }

    if (!connect->connectToken())
    {
        spdlog::error("{}:{} connect token is nullptr",
            LOG_FILE_PATH(__FILE__), __LINE__);
        return ErrCode_INVALID_ARGUMENT;
    }

    Token *token = reinterpret_cast<Token *>(connect->connectToken());
    try
    {
        m_stub = ff::QueueList::NewStub(token->channel);
        if (m_stub == nullptr)
        {
            spdlog::error("{}:{} Fail to get stub",
                LOG_FILE_PATH(__FILE__), __LINE__);
            return ErrCode_OS_ERROR;
        }
    }
    catch (...)
    {
        spdlog::error("{}:{} Fail to get stub", LOG_FILE_PATH(__FILE__), __LINE__);
        return ErrCode_OS_ERROR;
    }

    m_conn = connect;
    return ErrCode_OK;
}

u8 QueueList::createQueue(const std::string &name)
{
    spdlog::debug("{}:{} QueueList::createQueue",
        LOG_FILE_PATH(__FILE__), __LINE__);

    ff::QueueReq req;
    req.set_name(name);

    ff::Empty res;
    grpc::ClientContext ctx;

    Utils::setupCtx(ctx);
    grpc::Status status = m_stub->Create(&ctx, req, &res);
    if (status.ok())
    {
        return ErrCode_OK;
    }

    Utils::buildErrMsg(LOG_FILE_PATH(__FILE__), __LINE__, status);
    return ErrCode_OS_ERROR;
}

u8 QueueList::listQueue(std::vector<std::string> &out)
{
    spdlog::debug("{}:{} QueueList::listQueue",
        LOG_FILE_PATH(__FILE__), __LINE__);

    out.clear();
    out.reserve(100);

    ff::Empty req;
    ff::ListQueueRes res;
    grpc::ClientContext ctx;

    Utils::setupCtx(ctx);
    auto reader = m_stub->List(&ctx, req);
    if (reader == nullptr)
    {
        spdlog::error("{}:{} reader is nullptr", LOG_FILE_PATH(__FILE__), __LINE__);
        return ErrCode_OS_ERROR;
    }

    while (reader->Read(&res))
    {
        out.push_back(res.name());
    }

    grpc::Status status = reader->Finish();
    if (status.ok())
    {
        return ErrCode_OK;
    }

    Utils::buildErrMsg(LOG_FILE_PATH(__FILE__), __LINE__, status);
    return ErrCode_OS_ERROR;
}

u8 QueueList::deleteQueue(const std::string &name)
{
    spdlog::debug("{}:{} QueueList::deleteQueue",
        LOG_FILE_PATH(__FILE__), __LINE__);

    ff::QueueReq req;
    req.set_name(name);

    ff::Empty res;
    grpc::ClientContext ctx;

    Utils::setupCtx(ctx);
    grpc::Status status = m_stub->Delete(&ctx, req, &res);
    if (status.ok())
    {
        return ErrCode_OK;
    }

    Utils::buildErrMsg(LOG_FILE_PATH(__FILE__), __LINE__, status);
    return ErrCode_OS_ERROR;
}

u8 QueueList::renameQueue(const std::string &oldName,
                              const std::string &newName)
{
    spdlog::debug("{}:{} QueueList::renameQueue",
        LOG_FILE_PATH(__FILE__), __LINE__);

    ff::RenameQueueReq req;
    req.set_oldname(oldName);
    req.set_newname(newName);

    ff::Empty res;
    grpc::ClientContext ctx;

    Utils::setupCtx(ctx);
    grpc::Status status = m_stub->Rename(&ctx, req, &res);
    if (status.ok())
    {
        return ErrCode_OK;
    }

    Utils::buildErrMsg(LOG_FILE_PATH(__FILE__), __LINE__, status);
    return ErrCode_OS_ERROR;
}

std::shared_ptr<IQueue> QueueList::getQueue(const std::string &name)
{
    spdlog::debug("{}:{} QueueList::getQueue",
        LOG_FILE_PATH(__FILE__), __LINE__);

    ff::QueueReq req;
    req.set_name(name);

    ff::Empty res;
    grpc::ClientContext ctx;

    Utils::setupCtx(ctx);
    grpc::Status status = m_stub->GetQueue(&ctx, req, &res);
    if (status.ok())
    {
        Queue *queue = new (std::nothrow) Queue;
        if (!queue)
        {
            spdlog::error("{}:{} Fail to allocate memory",
                LOG_FILE_PATH(__FILE__), __LINE__);
            return nullptr;
        }

        if (queue->init(m_conn, nullptr, name))
        {
            spdlog::error("{}:{} Fail to initialize queue",
                LOG_FILE_PATH(__FILE__), __LINE__);
            delete queue;
            return nullptr;
        }

        return std::shared_ptr<IQueue>(queue);
    }

    Utils::buildErrMsg(LOG_FILE_PATH(__FILE__), __LINE__, status);
    return nullptr;
}

} // end namespace GRPC

} // end namespace DAO

} // end namespace Model
